core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -127,9 +128,10 @@ public TieredBlockStore() {
 
     BlockMetadataEvictorView initManagerView = new BlockMetadataEvictorView(mMetaManager,
         Collections.<Long>emptySet(), Collections.<Long>emptySet());
-    mAllocator = Allocator.Factory.create(initManagerView);
-    if (mAllocator instanceof BlockStoreEventListener) {
-      registerBlockStoreEventListener((BlockStoreEventListener) mAllocator);
+
+    mAllocationCoordinator = AllocationCoordinator.getInstance(initManagerView);
```

Comment:
`AllocationCoordinator.getInstance()`, formerly `Allocator.Factory.create(initManagerView)` is called twice in non-test code. This can cause issues where the Allocator is stateful like `RoundRobinAllocator`. Honestly I don't think anyone is using or should use `RoundRobinAllocator`, but anyways.

The 2 references are here and https://github.com/Alluxio/alluxio/blob/355bca2291caef1370858684316b6a188d4f105e/core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java#L119, see my change in `BlockMetadataManager`.

I think using a singleton can avoid having multiple stateful Allocator instances maintaining separate states. This is currently not a very critical issue yet, because currently we are creating 2 Allocator instances, both from empty block view in the same `TieredBlockStore` constructor. But I think this change may help in the future. What do you think @gpang ?

Reply:
It is not clear to me why the allocation coordinator needs to be a singleton? If the only times we create it is in the constructor of `TieredBlockStore` then it seems trivial to make it not a singleton and just use a single instance?

I also don't understand how a stateful allocator affects the singleton/instance decision, since the allocator state is living in the allocatorCoordinator, which is either a single instance or a singleton.